### DISCUSSION TITLE
Conversion of DWIN Implementation For Use On 0.3.2-DWTJ

### DISCUSSION THREAD

**Discussion by jgbrown54** [Guizz27/DWIN_LK4Pro#13] [General]
It turns out that the display board is a TJC4827X343 made by Techonics programmed using Nextion Editor. It is almost identical to the DWIN display but updated using a single .tft file. Before I get too far into this effort, I want to compare the Nextion Editor with whatever you use to program for the DWIN. What platform do you use for programming?

**Comment by Guizz27** [comment]
I use DGUS Tool version 7.381. You can find a more recent version 7.637 called DGUS_V7637 on this page https://www.dwin-global.com/tool/.
Version 8 of the same software will not work.

I believe even if you manage to build a screen firmware for the Nextion display, the code in Marlin will not work. Looking at Longer's source code, they actually have a separate class for TFT and DW.

**Comment by DanielJMorel** [comment]
are there any options for getting the TJC4827X343 up and running?  Your work with the LK4 is very impressive.  Nicely done!  I was hoping to upgrade my Longer LK5 Pro to an SKR 1.4 Turbo motherboard so that I could use Linear advance and Input Shaping but I can't get the touch screen to work with any of my Marlin compilations.  I am afraid if I mess with the original Longer LK5 Pro motherboard I will be left with an unusable printer.  Any help on this would be greatly appreciated.

**Comment by jgbrown54** [comment]
Since Longer's source code for Marlin is available, we could enable Linear Advance in their source and recompile it. That would work with the current Techonics display but it would not be the latest version of Marlin. Another option would be to purchase the DWIN display that is available. That would work with the advanced firmware. A third option would be to remove the display in favor of a browser based control panel. That would work with any motherboard or firmware. There is also an implementation of Linear Advance available using the CURA Ultimaker slicer that could be used.

**Reply by DanielJMorel** [reply]
I have purchased a BigTreeTech Pi TFT43 to go with the SKR 1.4 Turbo that I have.  I don't know if they are compatible with each other. 
 I guess i will find out.  I am hoping that it will work and that it will fit where the Longer LK5 Pro touchscreen was.  This way I don't have to take a chance on messing up my printer.  I can always fall back on the original set up if the Big Tree Tech items don't work out.  My apologies for being a newb to this.  Would this fit your browser based display suggestion?  Oh, by the way,  thank you for taking the time to reply.  It is much appreciated.  I am in way over my head!

**Comment by jgbrown54** [comment]
No problem. I help lots of people. Yes, My browser based control works on any firmware and includes a large display for PCs and Tablets and a small touchscreen display for Tablets and phones. I'm also working on a replacement touch screen display for Longer. See htps://Pixie-Server.com.

**Reply by DanielJMorel** [reply]
I tried to go to the link you provided but it gets a 404 This page is not the page you are looking for.

**Comment by jgbrown54** [comment]


**Reply by DanielJMorel** [reply]
That worked!  Thank you.  with this device it looks like I wouldn't need my raspberry Pi.  Nice work!  Looks very robust.  Way more powerful than the raspberry pi I am currently using.

**Comment by DanielJMorel** [comment]
My apologies for all the questions and thank you again for your help on this.  

Can you help me understand a little bit more about how Marlin and touch screen work?  I have a little experience in coding, just enough to get me in trouble.  Anyway, am I correct in thinking that Marlin includes some file that interacts with the touch screen.  If so, do you know what file and what include?  Looking at Longer’s configuration.h file it is hard to tell because of all the if statements for other printers.  Also, in your work on the LK4 you were able to customize the touch screen successfully but in that GitHub you mentioned not to use it if I showed “DWTJ” in the about screen.  My Longer has that.  Can you elaborate a little bit about why it can’t be used.  I’m guessing that this is a pretty involved setup to get it working so I will understand if you would prefer not to go into it.  

Do you always have to upgrade the touch screen when you update the firmware on the motherboard?  Are they reversible?  My current Mainboard firmware is 0.4.0-Marlin 2x and my Screen firmware is 0.3.1-DWTJ.  Are those the latest updates for both on the Longer LK5?  I am cautious about doing anything to my Longer motherboard in fear of not being able to return it to a working order.

Once again thank you for you help, it is very much appreciated.

Dan

**Comment by DanielJMorel** [comment]
Very impressive!  I wish I knew about this before I invested $250.00 in my Raspberry Pi.  It has a much better price point and it looks like it will provide much more control over the printer.  I may go that way and use the Pi for something else.  You should put a video out on Youtube so that others can know about this option.  When studying up on how to print without having to use the SD card, all I could find out there were reference to the Raspberry Pi.  I think that you Pixie has great potential.

**Comment by jgbrown54** [comment]
I'm sorry. I didn't mean to use your GitHub for advertising. But I thought it might help. Please disagree with what I'm about to say below if I am wrong.

A simple description of communication between Longer and the DWIN display would be to say it communicates with tokens. Longer sends a token saying which screen to display. Then when an area of the screen is pressed, the display sends a token to Longer and based on which screen Longer told DWIN to display, it determines what action to take for that token.  There is a little more to it than that. The screen can also jump to another screen on it's own when a area is touched and can even perform some simple calculations. To design for the DWIN display, you provide a graphic to be displayed complete with any icons or text representing actions from the screen. You then define areas on the screen that when touched, send a token to Longer or complete an action of their own. 

The Techonics display is similar in that it communicates with Longer using tokens. Except the format of the tokens is different. The screens are even programmed similarly by creating graphics and defining areas to be touched to send a token. I looked at trying to recreate the DWIN screens for the Techonics and ran into a major hurdle. The programming application for the Techonics display is only in Chinese. I did find an older version of the programmer that had been converted to English and I was hopeful that I could use it to create the screens. But the older programmer would not recognize the newer display and the newer display would not recognize TFT files created by the older programmer.

To add to that, the applications in Marlin that communicate with the DWIN display are now built into Marlin. (DGUS_RELOADED) But there are major limitations such as requiring a bed probe. And some other options. Since my overall goal was to provide enhanced firmware whatever the configuration was on your Longer printer, I tried to work my way through the code to modify it to remove those required options. After spending a couple of weeks on that project, I finally gave up. I would have success in making one thing work correctly but then it would break two other things. I then decided it would be easier to reach my goal by using an external platform, such as a PI, loaded with an existing application from a company in Germany called Repetier. I obtained a OEM license from them to keep the cost down and hardware from OrangePi and put together package, that can be implemented by a novice, that will not only allow communications with a Longer printer but with any printer with a USB interface no mater what firmware their printer runs, and all the controls are available from any browser connected to your network. Since you already have a Raspberry PI, you could load and run the Repetier software on it depending on what version of the Raspberry you have. If you are familiar with Octoprint, the Pixie is like Octoprint on steroids and can be implemented by a beginner if they can follow instructions.

In answer to your other questions, so far, the one screen set has worked with all version of Guizz27's enhanced firmware and yes you can always go back to Longer's firmware and their screen implementation.

**Comment by DanielJMorel** [comment]
No worries!  I  appreciate the heads up.  I will probably get a Pixie because of the robust features that it adds.  Very nice work!  I will use my Pi for learning and exploring Linux.  Thank you for telling me about it.

I can not thank you enough for your detailed explanation of how the DWIN and Marlin interact. and also thank for all the time you put into trying to make it work.  I would be very frustrated if I couldn’t get it to work.  I do follow what you are saying and I’m pretty sure that this is where I move on to another route.  I am able to connect with the SKR 1.4 Turbo through Octoprint and control the board.  As I said, I have ordered a Pi TFT43 which I’m hoping will work with the SKR 1.4 Turbo.  I do want to enable Input Shaping and Linear Advance and I should be able to do that with the TMC 2209 drivers.  I don’t care if I enable senseless homing but I’ll take it if it works.

Thanks again for all you work.  Very Much appreciated

Dan

**Comment by Guizz27** [comment]
Just to add to what James said about Marlin and DWIN.

First, the code in Longer's official firmware is honestly a mess. They added code a little bit everywhere in Marlin making it difficult to modify or integrate newer Marlin versions.

The code I use for this firmware was written originally by Desuuuu which is the author of the dgus_reloaded which was added to Marlin later. My version of dgus_reloaded is slightly different from the original like James mentioned, my version does not require a BLTouch, it behaves better for pausing (M25) and supports M600 for filament change. It also has more features available on the screen like e-steps, or setting K value for linear advance. I also have a few modifications elsewhere in Marlin's code for better pause support.

I version the screen and the printer firmware with "GC X.X". As long as the major version number is the same the screen and printer firware will be compatible. But none of these are compatible with Longer's firmware.

At the moment, creating the screen firmware is not an easy task and Techtronics looks even worse. 
DWIN (DW) screens require each display screens to be bitmaps. More recent versions allow Marlin to draw on the screen which means you can change the UI without having to build a new screen firmware.

But different DWIN screen versions (similar for other screens) need different DGUS software to create the firmware. You need the correct version of the software for the screen you are developing for. Using the later version of the software does not work with older screens for example. Also, half the software UI is in Chinese and I don't read Chinese. 

I'm still working on a screen firmware generator that would allow more features and customizations but I have to reverse engineer some of DWIN file formats.

I'm sure you will like your SKR board.

**Comment by DanielJMorel** [comment]
I can’t thank you guys enough for the information you’ve provided.  I have spent weeks trying to make this work and now I know that if you all are having a hard time making the Longer firmware(s) work then I am sure it is way past anything I can do with it.  I will stick to making the Big Tree Tech Motherboard and 4.3” screen work and leave the Longer Motherboard and screen in a box on a shelf and probably never look back.  I really appreciate	all the time you all have taken to bring me up to speed.  Thank you very much.

Dan
